Initial Android 12 updates (#495)
smartdevicelink/sdl_app_library_guides: docs/Migrating to Newer SDL Java Suite Versions/Updating to v5.4/index.md

# Upgrading to 5.4

## Overview

> This guide is to help developers get setup with the SDL Java library version 5.4. It is assumed that the developer is already updated to at least version 5.3 of the library.

Reviewer: ```suggestion
This guide is to help developers get set up with the SDL Java Suite library version 5.4. It is assumed that the developer is already updated to at least version 5.3 of the library.
```

The full release notes are published [here](https://github.com/smartdevicelink/sdl_java_suite/releases).

> SDL Java library version 5.4 adds support for Android 12.

Reviewer: ```suggestion
SDL Java Suite library version 5.4 adds support for Android 12.
```

## AndroidManifest Exported Flag

> Starting in Android 12, any activities, services, or broadcast receivers that use intent filters will need to explicitly declare the android:exported attribute for the given app components. The SdlRouterService and SdlReceiver should already have the exported attribute defined and set to true, but the USBAccessoryAttachmentActivity will now also require this attribute to be set. Any activity that had an intent-filter would have a default exported value of true. Now we need to explicitly set it.

Reviewer: ```suggestion
Starting in Android 12, any activities, services, or broadcast receivers that use intent filters will need to explicitly declare the `android:exported` attribute for the given app components. The SdlRouterService and SdlReceiver should already have the exported attribute defined and set to true, but the USBAccessoryAttachmentActivity will now also require this attribute to be set. Any activity that had an intent-filter would have a default exported value of true. Now we need to explicitly set it.
```

```xml
<activity
    android:name="com.smartdevicelink.transport.USBAccessoryAttachmentActivity"
    android:exported="true" <!--New Addition-->
    android:launchMode="singleTop">
    <intent-filter>
        <action android:name="android.hardware.usb.action.USB_ACCESSORY_ATTACHED" />
    </intent-filter>
    <!-- ... -->
</activity>
```

##Bluetooth Runtime Permissions

> Starting in Android 12 for the library to be able to connect to the HMI over bluetooth, App Developers will need to request the new `BLUETOOTH_CONNECT` runtime permission.

Reviewer: ```suggestion
Starting in Android 12, for the library to be able to connect to the HMI over Bluetooth, app developers will need to request the new `BLUETOOTH_CONNECT` runtime permission.
```

This means the permission will need to be listed in the `AndroidManifest.xml` file. The developer will also need to request this permission from the user as it is a runtime permission.

```xml
<uses-permission android:name="android.permission.BLUETOOTH_CONNECT"
    tools:targetApi="31"/>
```

## Foreground Services

Starting with Android 12, apps cannot start services from the background. In order to allow the library to work as intended, the library will now need to start the "SdlService" from the context of the active router service.

To achieve this there are a few changes that will be required in your application.

> First to allow your "SdlService" to be started from an external source (The active router service may belong to another app), you will need to export the service in your `AndroidManifest.xml`.

Reviewer: ```suggestion
First to allow your "SdlService" to be started from an external source (the active router service may belong to another app), you will need to export the service in your `AndroidManifest.xml`.
```

                    PendingIntent pendingIntent = (PendingIntent) intent.getParcelableExtra(TransportConstants.PENDING_INTENT_EXTRA);
                    try {
                        pendingIntent.send(context, 0, intent);
                    } catch (PendingIntent.CanceledException e) {
                        e.printStackTrace();
                    }
                }
            }
        } else {
            // SdlService needs to be foregrounded in Android O and above
            // This will prevent apps in the background from crashing when they try to start SdlService
            // Because Android O doesn't allow background apps to start background services
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.startForegroundService(intent);
            } else {
                context.startService(intent);
            }
        }
    }
//...
```

## SdlService Class Name

There is now an overridable method, `getSdlServiceName` in the `SdlBroadcastReceiver` class. This method is used by the `SdlBroadcastReceiver` to catch a possible foreground exception.

> When the app tries to start the `SdlService`, if the service does not enter the foreground within a set amount of time (This time is designated by the Android operating system) an exception will be thrown and the app may encounter and ANR.

Reviewer: ```suggestion
When the app tries to start the `SdlService`, if the service does not enter the foreground within a set amount of time (this time is designated by the Android operating system), an exception will be thrown and the app may encounter an ANR.
```

The `SdlBroadcasterReceiver` can catch this exception and prevent the ANR but will need to know the name of the class that throws the exception.

> By default the `getSdlServiceName` method will return "SdlService", if your app uses a name other than "SdlService" you will need to override `getSdlServiceName` in the `SdlReceiver` class to return the correct name.

Reviewer: ```suggestion
By default the `getSdlServiceName` method will return "SdlService". If your app uses a name other than "SdlService" you will need to override `getSdlServiceName` in the `SdlReceiver` class to return the correct name.
```

When the app tries to start the `SdlService`, if the service does not enter the foreground within a set amount of time (this time is designated by the Android operating system), an exception will be thrown and the app may encounter an ANR.

The `SdlBroadcasterReceiver` can catch this exception and prevent the ANR but will need to know the name of the class that throws the exception.

By default the `getSdlServiceName` method will return "SdlService". If your app uses a name other than "SdlService" you will need to override `getSdlServiceName` in the `SdlReceiver` class to return the correct name.

```java
//...
@Override
public String getSdlServiceName() {
    return SDL_SERVICE_CLASS_NAME;
}
//...
```

## Known Issue

When the user connects their device over USB and the user has not been granted Bluetooth Permissions the user will be presented a notification which will help navigate the user to grant Bluetooth Permissions for the app.

Once the permissions are granted the Router Service will open the Bluetooth connection.

> If the user then revokes these permissions, the Android operating system will kill Application running the Router Service and the Router Service process and none of the services callbacks will be called. Even though the Router Service has been killed the apps will still appear on the HMI.

Reviewer 2: ```suggestion
If the user then revokes these permissions, the Android operating system will kill the application running the Router Service and the Router Service process. Additionally, none of the service's callbacks will be called. Even though the Router Service has been killed, the HMI will still show the previously connected apps.
```

    return SDL_SERVICE_CLASS_NAME;
}
//...
```

## Known Issue

> When the user connects their device over USB and the user has not been granted Bluetooth Permissions the user will be presented a notification which will help navigate the user to grant Bluetooth Permissions for the app.

Reviewer: ```suggestion
When the user connects their device over USB and the user has not been granted Bluetooth Permissions the user will be presented with a notification which will help navigate the user to grant Bluetooth Permissions for the app.
```

Once the permissions are granted the Router Service will open the Bluetooth connection.

> If the user then revokes these permissions, the Android operating system will kill Application running the Router Service and the Router Service process and none of the services callbacks will be called. Even though the Router Service has been killed the apps will still appear on the HMI.

Reviewer: ```suggestion
If the user then revokes these permissions, the Android operating system will kill the application running the Router Service and the Router Service process, and none of the service's callbacks will be called. Even though the Router Service has been killed, the apps will still appear on the HMI.
```
